internal/utils/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,60 @@
+package ringbuffer
+
+type RingBuffer[T any] struct {
+	ring             []T
+	headPos, tailPos int
+	full             bool
+}
+
+func (r *RingBuffer[T]) Init(size int) {
+	r.ring = make([]T, size)
+}
+
+func (r *RingBuffer[T]) Len() int {
+	if r.full {
+		return len(r.ring)
+	}
+	if r.tailPos >= r.headPos {
+		return r.tailPos - r.headPos
+	}
+	return r.tailPos - r.headPos + len(r.ring)
+}
+
+func (r *RingBuffer[T]) PushBack(t T) {
+	if r.full {
+		r.grow()
+	}
+	r.ring[r.tailPos] = t
+	r.tailPos++
+	if r.tailPos == len(r.ring) {
+		r.tailPos = 0
+	}
+	if r.tailPos == r.headPos {
+		r.full = true
+	}
+}
+
+func (r *RingBuffer[T]) PopFront() T {
+	r.full = false
+	t := r.ring[r.headPos]
+	r.headPos++
```

Comment:
To be on the safe side, should we overwrite the existing element with the zero value?